### PR TITLE
Document limit pushdown in the dev guide

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

This PR also includes an introduction to different types of pushdowns. It repeats most of the info from the Javadoc comments in the `ConnectorMetadata` interface, but it also mentions how additional information can be passed between the metadata service and record set provider inside the table handle. It also describes how to properly handle "impedance mismatch", where the data source might not support all order combinations, and to return an actionable error message.

> Is this change a fix, improvement, new feature, refactoring, or other?
improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
docs

> How would you describe this change to a non-technical end user or system administrator?
n/a

## Related issues, pull requests, and links

* Related to #12956 which describes other connector features that can be implemented in the metadata service

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: docs/src/main/sphinx/develop/connectors.rst [hunk 1]
```diff
@@ -79,33 +79,102 @@ features, like:
 Note that data modification also requires implementing
 a :ref:`connector-page-sink-provider`.
 
+When a ``SELECT`` query is being executed, and it contains specific clauses,
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:82]: Probably add a heading for `Pushdown`.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 2]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+the engine can try pushing down some of them into the connector, by calling
+one of the following methods of the ``ConnectorMetadata`` service:
+
+* ``applyLimit`` for the ``LIMIT`` clause, or ``applyTopN`` when ``ORDER BY``
+  is also present
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:87]: nit

```suggestion
* ``applyLimit`` for the ``LIMIT`` clause, or ``applyTopN`` when both ``LIMIT`` and ``ORDER BY``
  clauses are present
```

File: docs/src/main/sphinx/develop/connectors.rst [hunk 3]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+* ``applyFilter`` for the ``WHERE`` clause
+* ``applyProjection`` for expressions in the ``SELECT`` clause
+* ``applySample`` for the ``TABLESAMPLE`` clause
+* ``applyAggregation`` for the ``GROUP BY`` clause
+* ``applyJoin`` for the different types of the ``JOIN`` clause
+* ``applyTableFunction`` for the ``TABLE ()`` clause
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:93]: IIRC `TABLE ( )` is not PTF. It's just syntax sugar to wrap something in `SELECT * FROM`. e.g. try `TABLE orders` in trino-cli and it'll execute `SELECT * FROM orders`.

Or maybe I'm confused.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 4]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+* ``applyTableScanRedirect``
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:94]: missing description

File: docs/src/main/sphinx/develop/connectors.rst [hunk 5]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+
+
+Connectors can indicate whether they don't support a particular pushdown
+or that the action had no effect by returning ``Optional.empty()``. Connectors
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:98]:
```suggestion
Connectors can indicate that they don't support a particular pushdown
or that the action had no effect by returning ``Optional.empty()``. Connectors
```

File: docs/src/main/sphinx/develop/connectors.rst [hunk 6]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+
+  It's critical for connectors to return ``Optional.empty()`` if calling
+  this method has no effect for that invocation, even if the connector generally
+  supports a particular pushdown. Doing otherwise can cause the optimizer
+  to loop indefinitely.
+
+Otherwise, these methods return a result object containing the connectors
+table handle. The connector can store any kind of information in the table
+handle, as it will be passed later to other services it implements, like
+the record set or page source provider.
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:112]: nit: Use class names (or links to sections which explain them).

File: docs/src/main/sphinx/develop/connectors.rst [hunk 7]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+
+Limit and TopN pushdown
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When a ``SELECT`` query contains a :ref:`limit-clause`, the engine will try to
+push down the limit into the connector. If there's no ``ORDER BY`` clause,
+the ``applyLimit`` method of the connector metadata service will be called,
+and the connector can return results in an arbitrary order.
+Otherwise, the engine will call ``applyTopN``.
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:121]: Otherwise != `LIMIT + ORDER BY`. We can be more explicit.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 8]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+
+If the connector could benefit from the information but can't guarantee that it
+will be able to produce fewer rows than the provided limit, it should return
+a non-empty result containing a new handle for the derived table and
+the "limit guaranteed" or "topN guaranteed" flag set to false.
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:126]: using field names would probably make this more obvious.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 9]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+
+If the connector can guarantee it will produce fewer rows than the provided
+limit, it should return a non-empty result with the "limit guaranteed" or
+"topN guaranteed" flag set to true.
+
+.. note::
+
+  The ``applyTopN`` is the only method that receives sort items from the
+  ``ORDER BY`` clause. It will be called even if there's no ``LIMIT`` clause.
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:135]: This seems to conflict what we mentioned earlier in the list of `applyXXX` methods. We say `applyTopN` is called when query has both `ORDER BY` and `LIMIT` clauses. Here we say it is called even if no `LIMIT` is present.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 10]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+
+In an SQL query, the ``ORDER BY`` section can include any column with any order.
+But the data source for the connector might only support limited combinations.
+Plugin authors have to decide if the connector should ignore the pushdown,
+return all the data and let the engine sort it, or throw an exception
+to inform the user that particular order is not supported, if fetching all
+the data would be too expensive or time consuming. When throwing
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:142]: Do we have any connectors which already do this today?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 11]
```diff
@@ -79,33 +79,102 @@ features, like:
...
+an exception, use the ``TrinoException`` class with the ``INVALID_ORDER_BY``
+error code and an actionable message, to let users know how to write a valid
+query.
+
 .. _connector-split-manager:
 
 ConnectorSplitManager
-^^^^^^^^^^^^^^^^^^^^^
+---------------------
```
Comment (hashhar) [inline docs/src/main/sphinx/develop/connectors.rst:150]: Everything here and below could be separate commit.

**Review by hashhar — COMMENTED** [review]
Some comments. Looks good overall.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 1 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:82]: Can we add it when there's going to be at least one other type of pushdown documented?

File: docs/src/main/sphinx/develop/connectors.rst (hunk 2 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:87]: topN would be called when only `ORDER BY` is present

File: docs/src/main/sphinx/develop/connectors.rst (hunk 3 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:93]: I thought adding parenthesis would be enough, but I see how that can be confusing. BTW `TABLE <name>` doesn't work for subqueries:
```
trino:v2> table (select 1);
Query 20220725_063410_00002_jik9j failed: line 1:7: mismatched input '('. Expecting: <identifier>
table (select 1)
```

How about something like this?
```suggestion
* ``applyTableFunction`` for the ``TABLE ()`` table function clause
```

File: docs/src/main/sphinx/develop/connectors.rst (hunk 4 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:94]: I didn't know how to describe it. Any suggestions?

File: docs/src/main/sphinx/develop/connectors.rst (hunk 11 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:150]: It's related :-)

File: docs/src/main/sphinx/develop/connectors.rst (hunk 10 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:142]: I do, https://github.com/nineinchnick/trino-rest

File: docs/src/main/sphinx/develop/connectors.rst (hunk 9 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:135]: I fixed the earlier occurrences. This paragraph is correct.

**Comment by nineinchnick** [comment]
@hashhar thanks a lot! AC and PTAL

File: docs/src/main/sphinx/develop/connectors.rst (hunk 3 above)
Comment (hashhar) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks for explaining. Your suggestion definitely seems better.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 4 above)
Comment (hashhar) [reply docs/src/main/sphinx/develop/connectors.rst:94]: Maybe `to redirect table scans to a different connector`?

File: docs/src/main/sphinx/develop/connectors.rst (hunk 10 above)
Comment (hashhar) [reply docs/src/main/sphinx/develop/connectors.rst:143]: Thanks. However all error codes in StandardErrorCodes are supposed to be thrown by engine. Connectors should define their own. At-least that's my understanding from some previous PRs.

cc: @martint would know better.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 4 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks, done!

File: docs/src/main/sphinx/develop/connectors.rst (hunk 10 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:143]: Sure, I can describe how to create custom exceptions. I was planning on adding a whole paragraph/section about error handling but in another PR.

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:143]: @martint WDYT?

File: docs/src/main/sphinx/develop/connectors.rst [hunk 12]
```diff
@@ -79,33 +79,103 @@ features, like:
...
+
+* ``applyLimit`` for the ``LIMIT`` clause, if ``ORDER BY`` is not present,
+  or ``applyTopN`` when ``ORDER BY`` is present
+* ``applyFilter`` for the ``WHERE`` clause
+* ``applyProjection`` for expressions in the ``SELECT`` clause
+* ``applySample`` for the ``TABLESAMPLE`` clause
+* ``applyAggregation`` for the ``GROUP BY`` clause
+* ``applyJoin`` for the different types of the ``JOIN`` clause
+* ``applyTableFunction`` for the ``TABLE ()`` table function clause
+* ``applyTableScanRedirect`` to redirect table scans to a different connector
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:94]: The applyXXX methods may or may not correspond to SQL clauses. They are related to operations in the plan IR, not directly with SQL language features. For instance, both the HAVING and WHERE clause get implemented as a Filter in the IR. In some cases, filters can be synthesized from other SQL constructs (e.g., expression in the JOIN clause could be pushed down as filters and eventually make it to connectors via applyFilter).

Therefore, it's more correct and accurate to describe these operations in terms of what plan structures they apply to.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 13]
```diff
@@ -79,33 +79,103 @@ features, like:
...
+the ``ConnectorRecordSetProvider`` or ``ConnectorPageSourceProvider``.
+
+Limit and TopN pushdown
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When a ``SELECT`` query contains a :ref:`limit-clause`, the engine will try to
+push down the limit into the connector. If there's no ``ORDER BY`` clause,
+the ``applyLimit`` method of the connector metadata service will be called,
+and the connector can return results in an arbitrary order.
+When there's a ``ORDER BY`` clause present, the engine will call ``applyTopN``.
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:121]: This needs to be described in terms of IR elements, not SQL-level constructs

File: docs/src/main/sphinx/develop/connectors.rst (hunk 10 above)
Comment (martint) [reply docs/src/main/sphinx/develop/connectors.rst:143]: > However all error codes in StandardErrorCodes are supposed to be thrown by engine

If an error is covered by one of the entries in StandardErrorCodes, connectors should use it. They should only define custom codes for non-standard (i.e., connector-specific) errors.

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:143]: So I'll leave it as is. We'll have more opportunities to mention connector-specific errors.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 12 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: I reworded this section to mention the IR. I also removed these example clauses, so we could explain them better later in subsections as we do here for limit and top-N.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 13 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Can you give one example? I can only think about something like `When the plan contains operations related to a ... clause` but it seems too verbose to use too often.

Comment (martint) [reply docs/src/main/sphinx/develop/connectors.rst]: E.g.,

> When the plan contains a `Sort` operation followed by a `Limit` operation, the engine will call `applyTopN`

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thank you so much. I redacted this section. PTAL as it still might be phrased better, but I now feel more confident in writing about pushdowns.

**Comment by nineinchnick** [comment]
@martint @mosabua PTAL, I think I addressed all comments and this should be ready to go

**Comment by nineinchnick** [comment]
@martint PTAL

File: docs/src/main/sphinx/develop/connectors.rst [hunk 14]
```diff
@@ -223,10 +223,79 @@ features, like:
...
+a given query.
+
+.. warning::
+
+  It's critical for connectors to return ``Optional.empty()`` if calling
+  this method has no effect for that invocation, even if the connector generally
+  supports a particular pushdown. Doing otherwise can cause the optimizer
+  to loop indefinitely.
+
+Otherwise, these methods return a result object containing the connectors
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:253]: > the connectors table handle

It's actually a new table handle representing the virtual table derived from applying the operation (filter, project, limit, etc) to the table produced by the table scan node.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 15]
```diff
@@ -223,10 +223,79 @@ features, like:
...
+the ``ConnectorRecordSetProvider`` or ``ConnectorPageSourceProvider``.
+
+Limit and top-N pushdown
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+When executing a ``SELECT`` query with ``LIMIT`` or ``ORDER BY`` clauses,
+the query plan may contain a ``Sort`` or ``Limit`` operations.
+
+When the plan contains a ``Sort`` operation, the engine
+tries to push down the limit into the connector by calling the ``applyTopN``
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:265]: This is only when there is a Sort *and* a Limit (they first get converted to a TopN IR operation). A Sort alone would not result in a call to applyTopN

File: docs/src/main/sphinx/develop/connectors.rst [hunk 16]
```diff
@@ -223,10 +223,79 @@ features, like:
...
+``topNGuaranteed`` (in ``TopNApplicationResult``) flag set to false.
+
+If the connector can guarantee to produce fewer rows than the provided
+limit, it should return a non-empty result with the "limit guaranteed" or
+"topN guaranteed" flag set to true.
+
+.. note::
+
+  The ``applyTopN`` is the only method that receives sort items from the
+  ``Sort`` operation. It's called even if there's no ``Limit`` operation.
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:283]: I'm not sure I understand what this is trying to convey.

File: docs/src/main/sphinx/develop/connectors.rst [hunk 17]
```diff
@@ -238,17 +307,21 @@ the strategy employed by the Example HTTP connector.
...
 
-Given a split and a list of columns, the record set provider is
+Given a split, a table handle, and a list of columns, the record set provider is
 responsible for delivering data to the Trino execution engine.
-It creates a ``RecordSet``, which in turn creates a ``RecordCursor``
-that's used by Trino to read the column values for each row.
+
+It creates a ``RecordSet``, which in turn creates a ``RecordCursor`` that's used
+by Trino to read the column values for each row. It can use any additional
+information stored in the table handle object by methods the metadata service
```
Comment (martint) [inline docs/src/main/sphinx/develop/connectors.rst:317]: The information can be stored in the table handle or elsewhere. Conceptually, a "table handle" is just an opaque handle to a virtual table. It could be implemented as an id into a remote distributed store that keeps the details, or kept inline inside the handle itself.

It's better to think of the table handle as generalized identifier or key.

File: docs/src/main/sphinx/develop/connectors.rst (hunk 15 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:265]: How does that map to a query like `SELECT * FROM a ORDER BY b`? Would there be a `Limit` operation with a max value or it wouldn't push down the order at all?

Comment (hashhar) [reply docs/src/main/sphinx/develop/connectors.rst:265]: Order won't be pushed.

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:265]: Thanks. I can't explain why I thought this works differently. I checked the tests in my custom connectors and I couldn't find any with only `ORDER BY`. I guess this is why we need to have this in the developer guide :-)

File: docs/src/main/sphinx/develop/connectors.rst (hunk 16 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:283]: I removed the last sentence. I found it surprising that there's no `applySort` and I did not expect `applyTopN` to be the only way of pushing down ordering. Do you think it's worth keeping this note?

File: docs/src/main/sphinx/develop/connectors.rst (hunk 17 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst]: Thanks! I reworded it.

**Comment by nineinchnick** [comment]
@martint I addressed all comments, PTAL
